src/extract.js

```diff
@@ -1,74 +1,54 @@
-            result = combineResults(result, ...args);
-            // So we don't add identifiers like "if"
-            if (node.callee.type !== "Identifier") {
-                result = combineResults(result, extract(node.callee));
-            }
-            return result;
-        case "MemberExpression":
-            const objectChildren = extract(node.object);
-            // Only recurse on the property if it is not an identifier
-            const propertyChildren = node.property.type === "Identifier" ? {variables: [], functions: []} : extract(node.property);
-            return combineResults(objectChildren, propertyChildren);
-        // Rest of the cases contain a single variable
-        case "ThisExpression":
-        case "Identifier":
-        default:
-            return {
-                variables: [node],
-                functions: []
-            };
+export default function extractIdentifiers(node, {filter, addParents} = {}) {
+    if (addParents) {
+        setAll(node);
     }
-}
 
-function combineResults(...results) {
-    return results.reduce((acc, cur) => {
-        return {
-            variables: acc.variables.concat(cur.variables),
-            functions: acc.functions.concat(cur.functions)
-        };
-    }, {
-        variables: [],
-        functions: []
-    });
+    function _extractIdentifiers(node) {
+        switch(node.type) {
+            case "Literal":
+                return []
+            case "UnaryExpression":
+                return _extractIdentifiers(node.argument);
+            case "BinaryExpression":
+            case "LogicalExpression":
+                const {left, right} = node;
+                return [left, right].flatMap(_extractIdentifiers);
+            case "ConditionalExpression":
+                const {test, consequent, alternate} = node;
+                return [test, consequent, alternate].flatMap(_extractIdentifiers);
+            case "Compound":
+                return node.body.flatMap(_extractIdentifiers);
+            case "ArrayExpression":
+                return node.elements.flatMap(_extractIdentifiers);
+            case "CallExpression":
+                return [node.callee, ...node.arguments].flatMap(_extractIdentifiers);
+            case "MemberExpression":
+                const {object, property} = node;
+                const propertyChildren = property.type === "Identifier" ? [] : _extractIdentifiers(property);
+                return _extractIdentifiers(object).concat(propertyChildren);
+            // Rest of the cases contain a single variable
+            // Also check for filter condition
+            case "ThisExpression":
+            case "Identifier":
+            default:
+                const result = [];
+                if (!filter || (filter && filter(node))) {
+                    result.push(node);
+                }
+                return result;
+        }
+    }
+
+    return _extractIdentifiers(node);
```

Comment:
Why do we need a separate function here? It seems to take the same arguments?
Other functions use this pattern when they need to pass *more* arguments to the recursive function.

Reply:
I did this because it was easier to only have to pass in 1 argument in recursive subcalls rather than all of them. Plus, then `parents.setAll` only needs to be called once at the start of the function (I know that even if it gets called multiple times it will only walk the tree once because it uses the heuristic of if a node has a parent pointer then it doesn't walk its subtree, but still)

If you don't like this and would prefer it be simpler, I can unnest it